File: array/186.cpp
#include <bits/stdc++.h>

using namespace std;

using vi = vector<int>;

void next_permutation(vector<int> &nums) {
    int largestIndex = -1;
    bool valid = false;

    for (int i = 1; i < nums.size(); ++i) {
        if (nums[i] > nums[i - 1] && i > largestIndex) {
            largestIndex = i;
            valid = true;
        }
    }

    if (valid == false) {
        reverse(nums.begin(), nums.end());
        return;
    }

    int largestSecondIndex;
    for (int j = largestIndex; j < nums.size(); ++j) {
        if (nums[j] > nums[largestIndex - 1]) largestSecondIndex = j;
    }

    swap(nums[largestSecondIndex], nums[largestIndex - 1]);

    reverse(nums.begin() + largestIndex, nums.end());
}

int main() {
    // vi nums = {0, 1, 2, 5, 3, 3, 0};
    vi nums = {1, 2, 3, 4, 5};
    for (int e : nums) cout << e << " ";

    cout << "\n";

    ::next_permutation(nums);

    for (int e : nums) cout << e << " ";

    return 0;
}

File: array/186.py
from typing import List


def nextPermutation(nums: List[int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """

    i, j = -1, -1
    valid = False

    for index in range(1, len(nums)):
        if nums[index] > nums[index - 1] and index > i:
            i = index
            valid = True

    if valid is False:
        nums.reverse()
        return

    for item in range(i, len(nums)):
        if nums[item] > nums[i - 1]:
            j = item

    nums[j], nums[i - 1] = nums[i - 1], nums[j]

    nums[i:] = reversed(nums[i:])


nums = [5, 4, 3, 2, 1]
nextPermutation(nums)
print(nums)
